src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+        // List plugins with same organisation (if known), name and version
+        Collection<PluginDescriptor> plugins = list(new Filter() {
+            @Override
+            public boolean accept(PluginDescriptor plugin) {
+                boolean match = true;
+                if ((organisation != null) && (plugin.organisation() != null)) {
+                    match = StringUtils.equalsIgnoreCase(organisation, plugin.organisation());
+                }
+                return match
+                        && StringUtils.equalsIgnoreCase(name, plugin.name())
+                        && StringUtils.equalsIgnoreCase(version, plugin.version());
+            }
+        });
+
+        // The same plugin can't exist multiple times... if so, reports the error
+        // because a manual clean up will be necessary
+        if ((plugins != null) && (!plugins.isEmpty())) {
+            if (plugins.size() > 1) {
+                throw new PluginRepositoryException(name(), "Plugin [organisation=" + organisation
+                        + ", name=" + name + ", version=" + version + "] exists in multiple instances");
+            }
+
+            return plugins.iterator().next();
+        }
+        return null;
+    }
+
+    @Override
+    public Collection<PluginDescriptor> list() {
+        return list(new Filter() {
+            @Override
+            public boolean accept(PluginDescriptor plugin) {
+                // Accepts all plugins
+                return true;
+            }
+        });
+    }
+
+    @Override
+    public Collection<PluginDescriptor> list(final Filter filter) {
+        // Get the es-plugin.properties filename
+        String esPluginProps = PluginsService.ES_PLUGIN_PROPERTIES;
+        if (settings != null) {
+            esPluginProps = settings.get(PluginsService.ES_PLUGIN_PROPERTIES_FILE_KEY, PluginsService.ES_PLUGIN_PROPERTIES);
+        }
+
+        final ImmutableList.Builder<PluginDescriptor> plugins = ImmutableList.builder();
+
+        String pluginName;
+        PluginDescriptor.Builder descriptor;
+
+        if (Files.exists(pluginsDir)) {
+            try (DirectoryStream<Path> root = Files.newDirectoryStream(pluginsDir)) {
+                for (Path pluginDir : root) {
+                    if (Files.isDirectory(pluginDir)) {
+
+                        // Determine the name of the plugin from the directory's name
+                        pluginName = pluginDir.getFileName().toString();
+                        descriptor = pluginDescriptor(pluginName);
+
+                        // Introspect JAR files to find plugin's description
+                        try (DirectoryStream<Path> jarFiles = Files.newDirectoryStream(pluginDir, "*" + JAR)) {
+                            for (Path jarFile : jarFiles) {
+                                // Indicates that the plugin has code/library
+                                descriptor.jvm();
```

Comment:
I'm thinking loud here. Wondering if we should be able to set or unset `jvm` property here, so `jvm(boolean)` although we don't need that today?